trie/patriciaMerkleTrie.go

```diff
@@ -691,6 +694,39 @@ func (tr *patriciaMerkleTrie) GetTrieStats(address string, rootHash []byte) (*st
 	return ts.GetTrieStats(), nil
 }
 
+// CollectLeavesForMigration will collect trie leaves that need to be migrated. The leaves are collected in the trieMigrator.
+// The traversing of the trie is done in a DFS manner, and it will stop when the gas runs out (this will be signaled by the trieMigrator).
+func (tr *patriciaMerkleTrie) CollectLeavesForMigration(
+	oldVersion core.TrieNodeVersion,
+	newVersion core.TrieNodeVersion,
+	trieMigrator vmcommon.DataTrieMigrator,
+) error {
+	tr.mutOperation.Lock()
+	defer tr.mutOperation.Unlock()
+
+	if check.IfNil(tr.root) {
+		return nil
+	}
+	if check.IfNil(trieMigrator) {
+		return errors.ErrNilTrieMigrator
+	}
+
+	if newVersion > core.MaxValidTrieNodeVersion || oldVersion > core.MaxValidTrieNodeVersion {
+		return fmt.Errorf("%w: newVersion %v,  oldVersion %v", errors.ErrInvalidTrieNodeVersion, newVersion, oldVersion)
+	}
+
+	if newVersion == core.NotSpecified && oldVersion == core.AutoBalanceEnabled {
```

Comment:
this can be a function that will handle all the incompatible values that we might have.

Reply:
Extracted all the version checks in a new func.